aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -434,6 +447,7 @@ void THTensor_(getRNGState)(THGenerator *_generator, THTensor *self)
   THArgCheck(THTensor_(isContiguous)(self), 1, "RNG state needs to be contiguous");
   rng_state = (THGenerator *)THTensor_(data)(self);
   THGenerator_copy(rng_state, _generator);
+  rng_state->mutex = NULL;  // mutex should not be part of the generator state
```

Comment:
This looks like it leaks the mutex. THGenerator_copy creates a new mutex, but then it's set to NULL here.

I think to do this properly, we should separate the RNG state from the rest of the THGenerator. For example, you can add a struct `THGeneratorState` which contains all the current fields of `THGenerator` except the `std::mutex*`. `THGenerator` can either extend `THGeneratorState` or contain it as a field. `THTensor_(getRNGState)` and `THTensor_(setRNGState)` should only access and overwrite the `THGeneratorState`.

Reply:
We still shouldn't be copying raw memory into a tensor. A correct solution would be to do it like we do in THC - copy the RNG state, then append the remaining fields with memcpy.

Reply:
What do you mean by "remaining fields"?

Reply:
Generator state (in the current sense) includes not only the mersene twister state, but also some other fields from THGenerator.

Reply:
Here is the definition of `THGenerator`. It seems to me that the state returned by `torch.get_rng_state()` should return exactly these fields (and not any mutex). These fields are all POD.

```
typedef struct THGenerator {
  /* The initial seed. */
  uint64_t the_initial_seed;
  int left;  /* = 1; */
  int seeded; /* = 0; */
  uint64_t next;
  uint64_t state[_MERSENNE_STATE_N]; /* the array for the state vector  */
  /********************************/

  /* For normal distribution */
  double normal_x;
  double normal_y;
  double normal_rho;
  int normal_is_valid; /* = 0; */
} THGenerator;
```

Reply:
They are POD, but POD doesn't mean that the struct doesn't have padding. We shouldn't be dumping padding bytes, nor should we depend on the actual layout of THGenerator in memory.

Reply:
There isn't padding in this structure. (All the fields are self-aligned). Even if there were, that's unrelated to this PR.